graphqlmetrics/pkg/batch/processor.go

```diff
@@ -0,0 +1,180 @@
+	shutdownChan chan struct{}
+	// Config is the configuration for the batch processor.
+	config ProcessorConfig
+
+	// queue is the channel to receive batches of buffer.
+	queue chan T
+
+	// buffer holds the buffered items which will be sent in a batch.
+	buffer []T
+
+	processBatch func([]T) error
+
+	// costFunc is a custom provided function, which will calculate a delta for
+	// each element that is added to the buffer.
+	costFunc func(T) int
+
+	// currentBufferCost holds the current total cost of the buffer.
+	currentBufferCost int
+}
+
+const defaultBufferSize = 10000
+
+// NewProcessor creates a new batch processor for a given configuration.
+// The ProcessFunc is the function to be invoked once a batch is ready.
+func NewProcessor[T any](
+	logger *zap.Logger,
+	config ProcessorConfig,
+	processFunc func([]T) error,
+	costFunc func(T) int,
+) *Processor[T] {
+	return &Processor[T]{
+		logger:       logger,
+		closeChan:    make(chan struct{}),
+		shutdownChan: make(chan struct{}),
+		config:       config,
+
+		queue:  make(chan T, config.MaxQueueSize),
+		buffer: make([]T, 0, defaultBufferSize),
+
+		processBatch: processFunc,
+		costFunc:     costFunc,
+	}
+}
+
+// Enqueue attempts to one or more elements to the queue. If the queue is full,
+// it will wait until the queue can accept more data or the context is cancelled.
+func (p *Processor[T]) Enqueue(ctx context.Context, element ...T) error {
+	for _, e := range element {
+		if !p.canAcceptBatch() {
+			return errors.New("processor is closed")
+		}
+
+		select {
+		case p.queue <- e:
+			p.logger.Debug("element added to the queue", zap.Any("element", e))
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return nil
+}
+
+// Start starts the batch processor.
+func (p *Processor[T]) Start(ctx context.Context) {
+	p.logger.Debug("Starting processor")
+
+	ticker := time.NewTicker(p.config.Interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			p.logger.Debug("interval completed - processing batch")
+			p.process()
+		case element := <-p.queue:
+			p.logger.Debug("receiving element from queue",
+				zap.Any("element", element),
+				zap.Int("queue_size", len(p.queue)),
+			)
+
+			p.receiveElement(element)
+		case <-ctx.Done():
+			p.process()
+			p.Stop()
+		case <-p.closeChan:
+			close(p.shutdownChan)
+			return
+		}
+	}
+}
+
+// Stop instructs to processor to shut down.
+// It will ensure that all remaining items in the queue and buffer are processed.
+func (p *Processor[T]) Stop() {
+	p.logger.Info("Shutdown requested - stopping processor")
+	close(p.closeChan)
+	close(p.queue)
+
+	<-p.shutdownChan
+
+	p.logger.Info("processing remaining items in the queue", zap.Int("queue_size", len(p.queue)))
+	for batch := range p.queue {
+		p.receiveElement(batch)
+	}
+
+	p.logger.Info("processing remaining items in the buffer", zap.Int("buffer_size", len(p.buffer)))
+	if len(p.buffer) > 0 {
+		p.process()
+	}
+}
+
+func (p *Processor[T]) canAcceptBatch() bool {
```

Comment:
How does it determine that the processor is ready to accept when it always return true?

Reply:
I got it. It used to block new requests on shutdown.